Graph-Kiota software requirements specifications (#58)
microsoftgraph/msgraph-sdk-design: graph-kiota/kiota.md

### What is Kiota? 
Learn [Kiota](https://github.com/microsoft/kiota#readme)

### APIs that can use Kiota 
- Any HTTP API with an Open API specification (For generators). Not limited to Microsoft Graph.

### Kiota libraries

Common for all languagues:
- Kiota generator: Generates the request builders

> The generated code contains the APIClient class and the request builders. 

Reviewer: I'd rephrase to request builders for fluent API and models. Api Client is just a special request builder.

```
class APIClient {
    httpCore : HttpCore;

    getResourceById(): ResourceRequestBuilder;
}
```

Libraries for each language:

- Kiota abstractions

>     - This library contains only abstract classes and interfaces which can be implemented to connect the generated request builders with a client library that makes the call to the API.

Reviewer: ```suggestion
    - This library contains only abstract classes and interfaces which can be implemented to connect the generated request builders with a client library that makes the call to the API, handles serialization and authentication.
```

    [Kiota abstractions dotnet](https://github.com/microsoft/kiota/tree/main/abstractions/dotnet)
    [Kiota abstractions java](https://github.com/microsoft/kiota/tree/main/abstractions/java)
    [Kiota abstractions typescript](https://github.com/microsoft/kiota/tree/main/abstractions/typescript)

    - Contains the following:
        - HttpCore interface

>         ```

Reviewer: again, let's refer to https://github.com/microsoft/kiota/blob/main/docs/kiotaabstractions.md#http-core instead of duplicating this.

Author: Created this here to capture the description and use case of each send function? 

Can you please provide the use cases for each of the send request, especially for `sendNoResponseContentAsync`? 

We can merge the documentations after we have captured information. 

We should update the code documentation as well.

Reviewer: https://github.com/microsoft/kiota/blob/afc5e21da5c52a0c7fc683c4bb7c91825cdfffcd/abstractions/typescript/src/httpCore.ts#L41

        interface HttpCore{
            getSerializationWriterFactory(): SerializationWriterFactory;
   
            sendAsync<ModelType extends Parsable>(requestInfo: RequestInformation, type: new() => ModelType, responseHandler: ResponseHandler | undefined): Promise<ModelType>;
            sendCollectionAsync<ModelType extends Parsable>(requestInfo: RequestInformation, type: new() => ModelType, responseHandler: ResponseHandler | undefined): Promise<ModelType[]>;
            sendPrimitiveAsync<ResponseType>(requestInfo: RequestInformation, responseType: "string" | "number" | "boolean" | "Date" | "ReadableStream", responseHandler: ResponseHandler | undefined): Promise<ResponseType>;
            sendNoResponseContentAsync(requestInfo: RequestInformation, responseHandler: ResponseHandler | undefined): Promise<void>;
            enableBackingStore(backingStoreFactory?: BackingStoreFactory | undefined): void;
        }
        ````
        - Serialization interfaces

>         - Authentication: `IAuthProvider`, abstract classes- `anonymousAuthenticationProvider` and `BearerTokenAuthneticationProvider`

Reviewer: ```suggestion
        - Authentication: `IAuthenticationProvider`, abstract classes- `anonymousAuthenticationProvider` and `BearerTokenAuthneticationProvider`
```

        - Utils
        - ResponseHandler interface

>         - MiddlewareOption class

Reviewer: ```suggestion
        - RequestOption interface
```
Anticipating the rename

        - RequestInformation
        - HttpMethod Enum

- Kiota core
    - This is generic core client library (functionality or purpose similar to Microsoft Graph core libraries)
    [Kiota core dotnet](https://github.com/microsoft/kiota/tree/main/http/dotnet/httpclient)
    [Kiota core java](https://github.com/microsoft/kiota/tree/main/http/java/okhttp)
    [Kiota core typescript](https://github.com/microsoft/kiota/tree/main/http/typescript/fetch)
    - This library contains the following:
        -  HttpCore class implementing the abstraction HttpCore interface.
        -  HttpClientbuilder (.NET, Java) or HTTPClient class (TS) (processes and records the middleware chain used while making requests.)

>         -  Middlewares: AuthHandler, Retry, Redirect, Compression, Caching, Telemetry

Reviewer: ```suggestion
        -  Middlewares: RetryHandler, Redirect, Compression, Caching, Telemetry
```

    
```

> class HttpCore implements kiota.abstractions.HttpCore {

Reviewer: let's leave that on the Kiota repo and refer to it

    sendCollectionAsync<ModelType extends Parsable>(requestInfo: RequestInformation, type: new() => ModelType, responseHandler: ResponseHandler | undefined): Promise<ModelType[]>;
    sendPrimitiveAsync<ResponseType>(requestInfo: RequestInformation, responseType: "string" | "number" | "boolean" | "Date" | "ReadableStream", responseHandler: ResponseHandler | undefined): Promise<ResponseType>;
    sendNoResponseContentAsync(requestInfo: RequestInformation, responseHandler: ResponseHandler | undefined): Promise<void>;
    enableBackingStore(backingStoreFactory?: BackingStoreFactory | undefined): void;
}

class HttpClient{
    middlewares : Middlewares();
    send() //calls the MiddlewareChain to make a  request with the native HttpClient;
}
```

Kiota Generated API Client (This will be the GraphServiceClient)

```
class APIClient {

    httpCoreInstance : HttpCore

    SampleRequestBuilder(httpCoreInstance);
    ExampleRequestBuilder(httpCoreInstance);
}
```

- Kiota serializations 
    - This library contains implementation of the kiota abstractions serialization interfaces
    - Currently targets `JSON` format.
    - Other formats to consider:

>         - Multipart payload

Reviewer: we did explore the multi-part aspect with @andrueastman and decided to leave it for now. Only OneNote uses it to our knowledge.

>         - Base64 encoded string

Reviewer: the case where it's part of a json payload should be handled today already. If the payload is only base64 encoded, this needs to be reviewed.

        - text/plain
        - xml

- Kiota auth providers

### Example of how we use the Kiota libraries:

> 1. Clone the Kiota repo, build the project and navigate the path containing the `kiota.exe`

Reviewer: can we refer to this documentation, and improve it if required? https://github.com/microsoft/kiota/#generating-sdks

Reviewer 2: There are language specific pages here https://github.com/microsoft/kiota/tree/main/docs/requiredtools that show how to use the Kiota tooling.  There dotnet one has an end to end experience.  It would be great to do the same for JS/TS

2. Generate the request builder using the command:
```
kiota.exe -d ../msgraph-sdk-powershell/openApiDocs/v1.0/mail.yml --language csharp -o ../somepath -n namespaceprefix
```
3. The generation should contain the `APIClient` class and request builders.

5. Create a demo class to test the `APIClient`.

4. Import the `KiotaCore` library

> - [dotnet](https://github.com/microsoft/kiota/packages/826853)

Reviewer: ```suggestion
- [dotnet](https://github.com/microsoft/kiota/blob/main/docs/requiredtools/dotnet.md)
```

> - [TypeScript](https://github.com/microsoft/kiota/packages/827040)

Reviewer: ```suggestion
- [TypeScript](https://github.com/microsoft/kiota/blob/main/docs/requiredtools/typescript.md)
```

Example:
```
import { HttpCore } from '@microsoft/kiota-http-fetch';
```

5. Import the serialization library
- [dotnet](https://github.com/microsoft/kiota/packages/826855)
- [TypeScript](https://github.com/microsoft/kiota/packages/827041)

```
import {JsonParseNodeFactory, JsonSerializationWriterFactory} from '@microsoft/kiota-serialization-json';
```

Example:

6. Import the AuthProvider library or implement the IAuthProvider interface to create a custom AuthProvider class.

> Find a suitable Kiota Authentication library in https://github.com/microsoft/kiota/packages

Reviewer: ```suggestion
Find a suitable Kiota Authentication library in https://github.com/microsoft/kiota/tree/main/authentication/typescript/azure
```

We could add a rollup of all the authentication packages in the /authentication folder, and one by language in the /authentication/language folder if it makes things easier.

> 7. Create a HTTPCore instance:

Reviewer: ```suggestion
7. Create a RequestAdapter instance:
```

```
var httpCore = new HTTPCore(authProvider, JsonParseNodeFactory, JsonSerializationWriterFactory)
```

8. Create a APIClient instance:

```
var apiClient = new APIClient(httpCore);

> var response = apiClient.users("<ID>").get();

Reviewer: ```suggestion
var response = await apiClient.users("<ID>").get();
```

> var httpCore = new HTTPCore(authProvider, JsonParseNodeFactory, JsonSerializationWriterFactory)

Reviewer: ```suggestion
var requestAdapter = new FetchRequestAdapter(authProvider, JsonParseNodeFactory, JsonSerializationWriterFactory)
```

> var apiClient = new APIClient(httpCore);

Reviewer: ```suggestion
var apiClient = new APIClient(requestAdapter);
```
